### PR TITLE
[dtensor][debug] add operation tracing to comm_mode

### PR DESCRIPTION
**Pull request by sinhaanshul**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #129613
* #129602
* __->__ #129017

**Summary**
I have added an even more detailed module tracker that now includes the collective counts and operations that happen in each submodule making it easier for users to debug. The tracing now includes the operation's DTensor arguements' input shape and sharding. Like the module collective tracing, the user also has the option to log the tracing table to output.txt file. I have decided not to include the example output for transformer as it is too many lines. The expected output for the MLP_operation_tracing is shown below:

Screenshot 2024-06-25 at 3 33 16 PM

Screenshot 2024-06-25 at 3 33 45 PM

**Test Plan**
1. torchrun --standalone --nnodes=1 --nproc-per-node=4 torch/distributed/_tensor/examples/comm_mode_features_example.py -e MLP_operation_tracing
2. torchrun --standalone --nnodes=1 --nproc-per-node=4 torch/distributed/_tensor/examples/comm_mode_features_example.py -e transformer_operation_tracing

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

### CODE REVIEW

**Comment by XilunWu** [comment]
let's convert this PR to Draft since it's still WIP.

**Review by tianyu-l — COMMENTED** [review]
This looks very interesting! Maybe similarly we can print out the placements of module parameters if they are DTensor?

**Comment by wz337** [comment]
Curious of what is your decision of printing out the mesh for the DTensor.

File: torch/distributed/_tensor/debug/comm_mode.py [hunk 1]
```diff
@@ -201,13 +258,19 @@ def __exit__(self, *args):
...
         table = ansi_escape.sub("", self.generate_module_tracing_table())
 
         with open("output.txt", "w") as log_file:
             log_file.write(table)
 
+    def log_operation_tracing_table_to_file(self):
+        ansi_escape = re.compile(r"\x1B\[[0-?]*[ -/]*[@-~]")
+        table = ansi_escape.sub("", self.generate_operation_tracing_table())
+
+        with open("output.txt", "w") as log_file:
```
Comment (XilunWu) [inline torch/distributed/_tensor/debug/comm_mode.py:271]: let's make the filename an argument. also in `log_module_tracing_table_to_file`.

**Review by XilunWu — APPROVED** [review]
LGTM! This change is a good fundament for future extension. Please address the comment in future PR.

**Pull request closed without merging**